span.go

```diff
@@ -129,6 +139,34 @@ func (s *spanImpl) FinishWithOptions(opts opentracing.FinishOptions) {
 
 	s.Lock()
 	defer s.Unlock()
+
+	if s.numDroppedLogs > 0 {
+		// We dropped some log events, which means that we used Logs as a circular
+		// buffer. De-circularize it.
+		pos := s.numDroppedLogs % s.tracer.options.MaxLogsPerSpan
+		// The oldest event is on position pos. We rotate (circular shift) the
+		// buffer in place, moving the first pos elements to the back. This
+		// algorithm is described in:
+		// 	http://www.cplusplus.com/reference/algorithm/rotate
+		for first, middle, next := 0, pos, pos; first != middle; {
+			s.raw.Logs[first], s.raw.Logs[next] = s.raw.Logs[next], s.raw.Logs[first]
+			first++
+			next++
+			if next == len(s.raw.Logs) {
+				next = middle
+			} else if first == middle {
+				middle = next
+			}
+		}
+		// Replace the oldest event. This means we are effectively dropping one more
+		// event.
+		s.raw.Logs[0].Event = fmt.Sprintf("** dropped %d events **", s.numDroppedLogs+1)
+		s.raw.Logs[0].Payload = nil
+		// Let the timestamp be: this event will reflect the timestamp of the last
+		// dropped event.
+	}
+
+	// The MaxLogsPerSpan limit does not apply to BulkLogData.
```

Comment:
The purpose of limiting is to cap memory usage. If the caller is supplying a huge `BulkLogData`, they are already storing that in memory. That said, I don't feel strongly - we can make those go through the normal limiting logic just as well.